src/chain/chain-training.cc

```diff
@@ -54,9 +64,24 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
       numerator.Backward(xent_output_deriv);
     }
   }
+  
+  num_logprob_weighted += -opts.boost * supervision.num_sequences *
+                          supervision.frames_per_sequence;
+  CuMatrixBase<BaseFloat> *boosting_mask = nnet_output_deriv;
+  if (opts.boost != 0.0 && opts.hard_boost) {
+    boosting_mask = new CuMatrix<BaseFloat>(*nnet_output_deriv);
+    boosting_mask->ApplyHeaviside();
```

Comment:
OK. 
BTW even for the purpose of testing the idea, I think this copying (`new CuMatrix<BaseFloat>(*nnet_output_deriv)`) might fail due to limited memory. I remember I had memory allocation problems with batch-sizes > 128 when I was testing on-GPU numerator computation with **two** nnet_output_deriv matrices.
Anyway I will test it to see what happens.